modules/cache-common/src/main/java/org/opensearch/cache/common/tier/TieredSpilloverCache.java

```diff
@@ -69,254 +72,529 @@ public class TieredSpilloverCache<K, V> implements ICache<K, V> {
-            new CacheConfig.Builder<K, V>().setRemovalListener(onHeapRemovalListener)
-                .setKeyType(builder.cacheConfig.getKeyType())
-                .setValueType(builder.cacheConfig.getValueType())
-                .setSettings(builder.cacheConfig.getSettings())
-                .setWeigher(builder.cacheConfig.getWeigher())
-                .setDimensionNames(builder.cacheConfig.getDimensionNames())
-                .setMaxSizeInBytes(builder.cacheConfig.getMaxSizeInBytes())
-                .setExpireAfterAccess(builder.cacheConfig.getExpireAfterAccess())
-                .setClusterSettings(builder.cacheConfig.getClusterSettings())
-                .setStatsTrackingEnabled(false)
-                .build(),
-            builder.cacheType,
-            builder.cacheFactories
-
-        );
-        this.diskCache = builder.diskCacheFactory.create(
-            new CacheConfig.Builder<K, V>().setRemovalListener(onDiskRemovalListener)
-                .setKeyType(builder.cacheConfig.getKeyType())
-                .setValueType(builder.cacheConfig.getValueType())
-                .setSettings(builder.cacheConfig.getSettings())
-                .setWeigher(builder.cacheConfig.getWeigher())
-                .setKeySerializer(builder.cacheConfig.getKeySerializer())
-                .setValueSerializer(builder.cacheConfig.getValueSerializer())
-                .setDimensionNames(builder.cacheConfig.getDimensionNames())
-                .setStatsTrackingEnabled(false)
-                .build(),
-            builder.cacheType,
-            builder.cacheFactories
-        );
+        Objects.requireNonNull(builder.cacheConfig.getSettings(), "settings can't be null");
+        if (builder.numberOfSegments <= 0) {
+            throw new IllegalArgumentException(NUMBER_OF_SEGMENTS_ZERO_EXCEPTION_MESSAGE);
+        }
+        this.numberOfSegments = builder.numberOfSegments;
         Boolean isDiskCacheEnabled = DISK_CACHE_ENABLED_SETTING_MAP.get(builder.cacheType).get(builder.cacheConfig.getSettings());
-        LinkedHashMap<ICache<K, V>, TierInfo> cacheListMap = new LinkedHashMap<>();
-        cacheListMap.put(onHeapCache, new TierInfo(true, TIER_DIMENSION_VALUE_ON_HEAP));
-        cacheListMap.put(diskCache, new TierInfo(isDiskCacheEnabled, TIER_DIMENSION_VALUE_DISK));
-        this.caches = Collections.synchronizedMap(cacheListMap);
-
         this.dimensionNames = builder.cacheConfig.getDimensionNames();
         // Pass "tier" as the innermost dimension name, in addition to whatever dimensions are specified for the cache as a whole
         this.statsHolder = new TieredSpilloverCacheStatsHolder(dimensionNames, isDiskCacheEnabled);
+        this.tieredSpilloverCacheSegments = new TieredSpilloverCacheSegment[this.numberOfSegments];
+        for (int i = 0; i < numberOfSegments; i++) {
+            tieredSpilloverCacheSegments[i] = new TieredSpilloverCacheSegment<K, V>(builder, statsHolder, i + 1, this.numberOfSegments);
+        }
```

Comment:
Although not part of this PR, just curious why do we need settingsUpdateConsumer for below setting. As per my understand, it is anyways not dynamic, right?
```
builder.cacheConfig.getClusterSettings()
            .addSettingsUpdateConsumer(DISK_CACHE_ENABLED_SETTING_MAP.get(builder.cacheType), this::enableDisableDiskCache);
```

Reply:
This is dynamic, we use this to enable/disable disk cache dynamically.